frame/contracts/src/storage/meter.rs

```diff
@@ -398,54 +397,60 @@ where
 	T: Config,
 	E: Ext<T>,
 {
-	/// Charge `diff` from the meter.
+	/// Charges `diff` from the meter.
 	pub fn charge(&mut self, diff: &Diff) {
 		match &mut self.own_contribution {
 			Contribution::Alive(own) => *own = own.saturating_add(diff),
 			_ => panic!("Charge is never called after termination; qed"),
 		};
 	}
 
-	/// Charge from `origin` a storage deposit for contract instantiation.
+	/// Adds a deposit charge.
+	///
+	/// Use this method instead of [`Self::charge`] when the charge is not the result of a storage
+	/// change. This is the case when a `delegate_dependency` is added or removed, or when the
+	/// `code_hash` is updated. [`Self::charge`] cannot be used here because we keep track of the
+	/// deposit charge separately from the storage charge.
+	pub fn charge_deposit(&mut self, deposit_account: DepositAccount<T>, amount: DepositOf<T>) {
+		self.total_deposit = self.total_deposit.saturating_add(&amount);
+		self.charges.push(Charge { deposit_account, amount, terminated: false });
+	}
+
+	/// Charges from `origin` a storage deposit for contract instantiation.
 	///
 	/// This immediately transfers the balance in order to create the account.
 	pub fn charge_instantiate(
 		&mut self,
 		origin: &T::AccountId,
 		contract: &T::AccountId,
-		info: &mut ContractInfo<T>,
+		contract_info: &mut ContractInfo<T>,
+		code_info: &CodeInfo<T>,
 	) -> Result<DepositOf<T>, DispatchError> {
 		debug_assert!(self.is_alive());
-
 		let ed = Pallet::<T>::min_balance();
-		let mut deposit =
-			Diff { bytes_added: info.encoded_size() as u32, items_added: 1, ..Default::default() }
-				.update_contract::<T>(None);
-
-		// Instantiate needs to transfer at least the minimum balance in order to pull the
-		// deposit account into existence.
-		// We also add another `ed` here which goes to the contract's own account into existence.
-		deposit = deposit.max(Deposit::Charge(ed)).saturating_add(&Deposit::Charge(ed));
-		if deposit.charge_or_zero() > self.limit {
+
+		let deposit = contract_info.update_base_deposit(&code_info);
+		if deposit > self.limit {
 			return Err(<Error<T>>::StorageDepositLimitExhausted.into())
 		}
```

Comment:
Shouldn't this limit be checked before the info gets updated?

Reply:
well we won't store the changes if some error happens 
https://github.com/paritytech/substrate/blob/9627c4c97d5696af0f2ac63fa504a22e239841ff/frame/contracts/src/exec.rs#L975-L977